[migration-tools] Always release source version with updated release increment instead of last_version.

nuclearsandwich opened the pull request:
When performing the Rolling platform migration several repositories had bloom releases which were not committed to the distribution file at the time of the source ref.
As a result the `last_version` entries for those repositories did not match the release version in the source distribution during migration.
Since the migration tool was using the last_release in order to handle any releases that were not hard coded to a specifc version and/or release tag this created a split between what was performed by git-bloom-release during migration and what was configured in the resulting distribution file.

One example is the ros2_control repository which had a 2.2.0-1 release made in https://github.com/ros2-gbp/ros2_control-release/commit/7504ac3640529603ee80dae424b516caf1a53409 but which never received a rosdistro PR.
Thus the migration tool released 2.2.0-2 based on the last_release information in the tracks file but the source distribution was anticipating 2.1.0-2.

Of the changes in this commit I am most hesitant about the release_increment handling since this script will eventually generate and force push updated release artifacts and if in a similar situation a 2.1.0-2 was released but not published in the source distribution file its artifacts would be clobbered where they overlapped with the migration-tool's own 2.1.0-2.
However I don't have a method of resolving that situation without searching each release repository's artifacts for a full list of release increments associated with a version.

nuclearsandwich committed:
Always release source version and source increment + 1 instead of last_version.

When performing the Rolling platform migration several repositories had bloom releases which were not committed to the distribution file at the time of the source ref.
As a result the `last_version` entries for those repositories did not match the release version in the source distribution during migration.
Since the migration tool was using the last_release in order to handle any releases that were not hard coded to a specific version and/or release tag this created a split between what was performed by git-bloom-release during migration and what was configured in the resulting distribution file.

One example is the ros2_control repository which had a 2.2.0-1 release made in https://github.com/ros2-gbp/ros2_control-release/commit/7504ac3640529603ee80dae424b516caf1a53409 but which never received a rosdistro PR.
Thus the migration tool released 2.2.0-2 based on the last_release information in the tracks file but the source distribution was anticipating 2.1.0-2.

Of the changes in this commit I am most hesitant about the release_increment handling since this script will eventually generate and force push updated release artifacts and if in a similar situation a 2.1.0-2 was released but not published in the source distribution file its artifacts would be clobbered where they overlapped with the migration-tool's own 2.1.0-2.
However I don't have a method of resolving that situation without searching each release repository's artifacts for a full list of release increments associated with a version.

cottsay commented on migration-tools/migrate-rosdistro.py:
```diff
@@ -183,20 +183,24 @@ def write_tracks_file(tracks, commit_msg=None):
             write_tracks_file(tracks, f'Update {args.dest} track to release exactly last-released tag.')
 
+        # Update release increment for the upcoming release.
+        release_inc = str(int(source_inc) + 1)
```
Rather than overwriting the tag, I'd opt for a release inc which is "newer than both". It should have the same effect without any force pushing.

```suggestion
        release_inc = str(max(int(source_inc), int(dest_track['release_inc'])) + 1)
```

tfoote replied:
+1 in case there was a missed push somewhere.

nuclearsandwich replied:
There are still cases this solution wouldn't cover with multiple uncommitted releases.

Using ros2_control as an example again. If our source distribution was at version 2.1.0-1 and the release repository contained tagged but unpublished in rosdistro releases 2.1.0-2 and 2.2.0-1, then the migration tool would see the last release `2.2.0` and release increment `"1"` in the tracks file and create a release `2.1.0-2` from
```python
    release_inc = str(max(int("1"), int("1")) + 1)
```
which would clobber the upstream 2.1.0-2.

This is what led me to think that the only reliable way to prevent clobbering unreleased tags is to scan for all tags matching the target release version to pick a strictly greater release increment.

cottsay replied:
I see - good catch. Well it's still an improvement, no?

nuclearsandwich replied:
Without a doubt.

The question hidden in my comment was Should we take the improvement and continue the migration or push for a solution that handles the further edge case?

nuclearsandwich replied:
I don't think that there's any current repositories with that specific scenario so I would be inclined to take your improvement and proceed with the migration. But I'll add a comment documenting the additional edge case and try to revisit it before we pull this out again for Humble.

cottsay replied:
I agree with your analysis - we can't do better without scanning the tags, which is a bit more work and IMO sort-of reaches further into the repositories than I think the migration tooling should.

+1 to move forward.

nuclearsandwich replied:
Updated in d80eb24.

cottsay reviewed (approved):
I suggested a possible approach to avoid the force-push, but either way, I think this should work OK.

tfoote reviewed (approved):
lgtm with same comment as Scott

nuclearsandwich committed:
Increment the greater between the source distribution increment and current release increment.

Reduce the likelihood that the migration will clobber existing release
tags by incrementing the greater between the source distribution
increment and the current release increment in the Bloom track.
There are still situations where a release could be clobbered which has
been documented but left unhandled.

Co-authored-by: Scott K Logan <logans@cottsay.net>

nuclearsandwich commented:
Whoops I pinged for additional review requests just before seeing @cottsay's additional :+1:. I've incorporated the suggestion in d80eb24 and I am pretty confident this covers all outstanding affected repositories as is.
